shared/src/server/proxy_protocol.rs

```diff
@@ -0,0 +1,257 @@
+use super::error::{ServerError, ServerResult};
+pub use ppp::v2::Header as PPHeader;
+use std::convert::TryFrom;
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+const MIN_PROXY_PROTOCOL_HEADER_LEN: usize = 16;
+
+pub struct AcceptedConn<S: AsyncRead + AsyncWrite> {
+    inner: S,
+    proxy_protocol: Option<PPHeader<'static>>,
+    preread_buffer: Option<Vec<u8>>,
+}
+
+impl<S: AsyncRead + AsyncWrite + Sync> AcceptedConn<S> {
+    fn new(
+        inner: S,
+        proxy_protocol: Option<PPHeader<'static>>,
+        opt_preread_buffer: Option<Vec<u8>>,
+    ) -> Self {
+        // if the preread buffer is empty, we don't need to store it
+        let preread_buffer = match opt_preread_buffer.as_deref() {
+            Some(buf) if buf.len() == 0 => None,
+            Some(_) => opt_preread_buffer,
+            None => None,
```

Comment:
nitpick to reduce branches:

```suggestion
            Some(buf) if buf.len() > 0 => buf,
            _ => None,
```